receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,313 @@
+	if wrapper.originalPort == 0 || crashCount > healthyCrashCount {
+		err := wrapper.Shutdown(wrapper.context)
+		if err != nil {
+			return crashCount, fmt.Errorf("could not stop receiver associated to process, killing it")
+		}
+	}
+
+	return crashCount, nil
+}
+
+// assignNewRandomPort generates a new port, creates a new metrics receiver with that port and assigns it to the wrapper
+func (wrapper *prometheusReceiverWrapper) assignNewRandomPort(oldPort int) (int, error) {
+	var err error
+	newPort := generateRandomPort(oldPort)
+
+	// Assign the new port in the config
+	wrapper.receiverConfig.PrometheusConfig.ScrapeConfigs[0].ServiceDiscoveryConfig.StaticConfigs[0].Targets = []model.LabelSet{
+		{model.AddressLabel: model.LabelValue(fmt.Sprintf("localhost:%v", newPort))},
+	}
+
+	// Create new Prometheus receiver with new config and replace pointer to it in wrapper
+	factory := &prometheusreceiver.Factory{}
+	wrapper.prometheusReceiver, err = factory.CreateMetricsReceiver(wrapper.context, wrapper.logger, wrapper.receiverConfig, wrapper.consumer)
+	if err != nil {
+		return 0, fmt.Errorf("unable to create Prometheus receiver. Killing this process")
+	}
+
+	return newPort, nil
+}
+
+// fillPortPlaceholders will check if any of the strings in the process data have the {{port}} placeholder, and replace it if necessary
+func (wrapper *prometheusReceiverWrapper) fillPortPlaceholders(newPort int) *subprocessmanager.SubprocessConfig {
+	var port string
+	if wrapper.originalPort == 0 {
+		port = strconv.Itoa(newPort)
+	} else {
+		port = strconv.Itoa(wrapper.originalPort)
+	}
+
+	newConfig := *wrapper.subprocessConfig
+	// ReplaceAll runs much faster (about 5x according to my tests) than checking for a regex match, therefore no checks are made and ReplaceAll simply returns the original string if no match is found
+	// ReplaceAll is run on the original strings of the config, which remain unchanged
+	newConfig.Command = strings.ReplaceAll(wrapper.config.SubprocessConfig.Command, portTemplate, port)
+
+	for i, env := range wrapper.config.SubprocessConfig.Env {
+		newConfig.Env[i].Value = strings.ReplaceAll(env.Value, portTemplate, port)
+	}
+
+	return &newConfig
+}
+
+// generateRandomPort will try to generate a random port until it is different than the last one generated
+func generateRandomPort(lastPort int) int {
+	var newPort int
+	for {
+		newPort = random.Intn(maxPortRange-minPortRange) + minPortRange
+
+		// Generate another port if it's the same
+		if newPort == lastPort {
+			continue
+		}
+
+		// Test the port to see if it's available
+		portTest, err := net.Listen("tcp", fmt.Sprintf(":%v", newPort))
+		if err != nil {
+			continue
+		}
+
+		portTest.Close()
+		break
+	}
+	return newPort
+}
+
+// Shutdown stops the underlying Prometheus receiver.
+func (wrapper *prometheusReceiverWrapper) Shutdown(ctx context.Context) error {
+	return wrapper.prometheusReceiver.Shutdown(ctx)
```

Comment:
What about also shutting down the loop running in `manageProcess`?

Reply:
This is just to shutdown the actual receiver, and is called from within `manageProcess`! maybe I'm missing something here?

Reply:
This function (the one we are inside, not the one you call here) is the `Shutdown` function for the prometheus exec receiver. It should be responsible for shutting down everything that is going on in this receiver, ideally including the `manageProcess` goroutine. Otherwise that will continue running forever.

Reply:
Alright, makes sense. I'm investigating how I can do this.

Reply:
This is generally done using channels, and will require a little bit of a refactor. In your main for loop you can select on a channel similar to what is described here: https://tour.golang.org/concurrency/5.

At any point when your code blocks (when subprocess is running or when waiting for a restart delay), you want to block on a channel so you can kill the goroutine if `Shutdown` is called (similar to https://github.com/open-telemetry/opentelemetry-collector/blob/master/extension/fluentbitextension/process.go#L145).

- To block on a channel when the subprocess is running, you can do something like this: https://github.com/open-telemetry/opentelemetry-collector/blob/master/extension/fluentbitextension/process.go#L178
- To block on a channel when you are waiting for a restart delay, you can use [time.After](https://golang.org/pkg/time/#After) instead of `time.Sleep`